_posts/2019-07-09-javelin.md

```diff
@@ -0,0 +1,25 @@
+
+SPIRV-Cross shows up quite a bit in our performance profiles (e.g. on Dota2). It is also written in a way (arguably!) that makes weak guarantees about what the state of the translator is at any point. And while it moves fast (in terms of contributions), it defines for us today what we can and what we can not do in terms of using advanced backend features, e.g. tessellation, argument buffers, etc. It complicates our [build process](https://github.com/gfx-rs/gfx/issues/2520), becoming a pain point for developers and users alike.
+
+We think the time has come for us [to attack](https://github.com/gfx-rs/gfx/issues/2269) that last bastion of C++ code in gfx-rs. That is what the [Javelin](https://github.com/gfx-rs/javelin) project is about: "A SPIR that flies above the garden walls". It's a very complex piece of software and we have not made much progress on it yet. However, we feel that Rust, once again, is the best tool for the job of shader translation: it's about parsing, working with bytes and data structures, with an ability to do unit and fuzzy testing, and no external dependencies.
+
+This post is meant to communicate our vision and plans for this future development. Note that here we are only talking about the path of shader translation from SPIR-V to anything, we don't plan to have GLSL->SPIR-V in scope. The key advantages we hope to gain by implementing Javelin are:
+
+  1. Achieve a pure-Rust code base, which is easier to build and port to platforms, such as the Web, or Android
+  2. Make the shader translation faster, and its artifacts easier to cache and reuse
+  3. Make it easier to maintain and identify issues in the shader translation
+  4. Start working on SPIR-V validation and sanitization for safety and WebGPU
+  5. Unlimit our progress with advanced backend features:
+    - explore direct translation to LLVM, MSL, and DXIL (all of those are LLVM-IR variations)
+    - get Metal argument buffers and tessellation in a way that works best for us
+
+In terms of implementation, the first step is getting the proper tools to work with SPIR-V. We are currently looking into building the proper structured representation in [rspirv](https://github.com/google/rspirv). Once this is done, we'd start experimenting with simple shader outputs in the backends behind an optional feature. We are also planning to have a web playground to be able to convert the shaders online and investigate issues. The path forward is hard, and we would appreciate any help!
```

Comment:
We should probably clarify how a shader playground would relate to the existing efforts of http://shader-playground.timjones.io/